core/trino-main/src/main/java/io/trino/execution/buffer/BroadcastOutputBuffer.java

```diff
@@ -412,7 +407,7 @@ private void noMoreBuffers()
 
     private void checkFlushComplete()
     {
-        if (state.get() != FLUSHING && state.get() != NO_MORE_BUFFERS) {
```

Comment:
This probably should be `BufferState state = stateMachine.get`
and then you should perform check. Otherwise `state` could move from `NO_MORE_BUFFERS` to `FLUSHING` between `stateMachine.getState()` calls, which seem racy

Reply:
Yeah, it does seem weird. I also thought about that. I don't know exactly why it is implemented this way. At the end of the day I decided not to change it and keep the change as close to being mechanic as possible.

Reply:
It's been like that for a very long time. It doesn't seem to be likely that the implementation is incorrect. But I agree, it's super confusing. Let me add a commit that simplifies it.